### PR TITLE
Adopt the MSAL broker to talk to the OS for Microsoft auth

### LINKED ISSUE #229431: [MSAL] Adopt Broker flow in Microsoft Auth provider via MSAL-node
Docs here:
https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/extensions/msal-node-extensions#usage---brokering

I talked to @deepak1556 about this one and the design we want to go with is having this native node module (`@azure/msal-node-extensions`) brought in to _Core_ and have a worker/process that could process these requests. This way the ExtHost is not polluted with this native node module.

So the Microsoft Authentication extension:
1. will talk to the Core window
2. (optional) Core window talks to the Main Process
3. and the Core window or Main process talks to this other process/worker

This could be API.. or maybe some other mechanism. API probably makes the most sense. Maybe a require interceptor if those still exist in a post-ESM world?

### PR DESCRIPTION
**Pull request by TylerLeonhardt**
This adopts the `NativeBrokerPlugin` provided by `@azure/msal-node-extensions` to provide the ability to use auth state from the OS, and show native auth dialogs instead of going to the browser.

This has several pieces:
* The adoption of the broker in the microsoft-authentication extension:
  * Adding `NativeBrokerPlugin` to our PCAs
  * Using the proposed handle API to pass the native window handle down to MSAL calls (btw, this API will change in a follow up PR)
  * Adopting an AccountAccess layer to handle:
    * giving the user control of which accounts VS Code uses
    * an eventing layer so that auth state can be updated across multiple windows
* Getting the extension to build properly and only build what it really needs. This required several package.json/webpack hacks:
  * Use a fake keytar since we don't use the feature in `@azure/msal-node-extensions` that uses keytar
  * Use a fake dpapi layer since we don't use the feature in  `@azure/msal-node-extensions` that uses it
  * Ensure the msal runtime `.node` and `.dll` files are included in the bundle
* Get the VS Code build to allow a native node module in an extension: by having a list of native extensions that will be built in the "ci" part of the build - in other words when VS Code is building on the target platform

There are a couple of followups:
* Refactor the `handle` API to handle (heh) Auxiliary Windows https://github.com/microsoft/vscode/issues/233106
* Separate the call to `acquireTokenSilent` and `acquireTokenInteractive` and all the usage of this native node module into a separate process or maybe in Core... we'll see. Something to experiment with after we have something working. NEEDS FOLLOW UP ISSUE

Fixes https://github.com/microsoft/vscode/issues/229431

### CODE REVIEW

File: extensions/microsoft-authentication/extension.webpack.config.js
Comment (TylerLeonhardt) [inline extensions/microsoft-authentication/extension.webpack.config.js:44]: This can go away when this is released: https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/7412

File: build/lib/extensions.ts
Comment (deepak1556) [inline build/lib/extensions.ts:358]: Can `glob.sync` operate on `extensionFilter` so that double packaging is avoided. It also aligns with the fact we know the list of extensions that should act as source in the `packageLocalNativeExtensionsStream` case.

File: build/gulpfile.vscode.js
Comment (deepak1556) [inline build/gulpfile.vscode.js:516]: Why is this changed ? Don't we want `compileNativeExtensionsBuildTask` in this case ?

File: build/gulpfile.vscode.js
Comment (deepak1556) [inline build/gulpfile.vscode.js:448]: Does the binary artifacts already have proper copyright and versioning, if not this shouldn't be skipped.

**Review by deepak1556 — COMMENTED** [review]
Reviewed the build part. LGTM overall, minor nits

**Comment by deepak1556** [comment]
Running a full build for sanity https://monacotools.visualstudio.com/Monaco/_build/results?buildId=304750&view=results

File: build/gulpfile.vscode.js
Comment (TylerLeonhardt) [reply build/gulpfile.vscode.js:516]: `compileExtensionsBuildTask` builds everything so I didn't want to duplicate work

Comment (deepak1556) [reply build/gulpfile.vscode.js:516]: I see, if we change the logic per https://github.com/microsoft/vscode/pull/233739/files#r1839347402 then we can avoid this change as well. Basically, making `compileExtensionsBuildTask` and `compileNativeExtensionsBuildTask` mutually exclusive

File: build/lib/extensions.ts
Comment (TylerLeonhardt) [reply build/lib/extensions.ts]: I separated things out so now there are mutually exclusive tasks and also a big task that does it all (for non-ci tasks like localization).

Let me know what you think!

**Comment by TylerLeonhardt** [comment]
New sanity build after changes: https://monacotools.visualstudio.com/Monaco/_build/results?buildId=304898&view=results

File: build/gulpfile.reh.js
Comment (deepak1556) [inline build/gulpfile.reh.js:482]: Don't we want `compileExtensionsBuildTask` here ? I believe this task is only used for local build, so useful to build all targets ?

File: build/gulpfile.vscode.js
Comment (deepak1556) [inline build/gulpfile.vscode.js:505]:
```suggestion
			compileExtensionsBuildTask,
```

?

File: build/gulpfile.vscode.js
Comment (deepak1556) [reply build/gulpfile.vscode.js]: I like it, thanks!

File: build/gulpfile.reh.js
Comment (TylerLeonhardt) [reply build/gulpfile.reh.js:482]: It does build all targets. We build the non-native ones here, and then the native ones in the `serverTaskCI` which is included in this local build task at the bottom

File: build/gulpfile.vscode.js
Comment (TylerLeonhardt) [reply build/gulpfile.vscode.js:505]: With this suggestion, we will be building the native ones twice. I separated them out so that we didn't double build based on your original feedback.

The non-native ones are built here, and then the native ones are part of `vscodeTaskCI` which is added to the end of this local build task.

Am I misunderstanding something?

**Comment by TylerLeonhardt** [comment]
Sanity build here: https://monacotools.visualstudio.com/Monaco/_build/results?buildId=305103&view=results

File: build/darwin/create-universal-app.ts
Comment (deepak1556) [reply build/darwin/create-universal-app.ts:32]: This is fine to unblock this PR, I will follow-up on the TODO

File: build/gulpfile.reh.js
Comment (deepak1556) [reply build/gulpfile.reh.js:482]: Sorry missed it

File: build/gulpfile.vscode.js
Comment (deepak1556) [reply build/gulpfile.vscode.js:505]: Sorry missed the tasks from `vscodeTaskCI`.

File: build/gulpfile.vscode.web.js
Comment (deepak1556) [inline build/gulpfile.vscode.web.js:205]: Not going to block the PR on this, but is building the native extensions have an effect in the web pipeline. Maybe we should skip the native ones ?

**Review by deepak1556 — APPROVED** [review]
Thanks for cleaning up the extension tasks and adding the support to include native modules with them. 🚀

File: build/gulpfile.vscode.web.js
Comment (TylerLeonhardt) [reply build/gulpfile.vscode.web.js:205]: In Microsoft Auth's case, it has web code as well... so it actually does need to be built for web.

**Pull request merged**
